### PR TITLE
Add power method for general f^g in the AutoDiffBlock

### PR DESCRIPTION
**Pull request by totto82**
A power method where both f and g are ADB variables is added
using the general derivative rule
(f^g)' = f^g * ln(f) * g' + g * f^(g-1) * f'

Tests are added to test_block.cpp

### CODE REVIEW

**Comment by atgeirr** [comment]
I have to admit I had to work a bit to verify the formula, it's been a long time since calculus class and logarithmic differentiation! Having checked that, I'll look at the code.

File: opm/autodiff/AutoDiffBlock.hpp [hunk 1]
```diff
@@ -676,6 +677,50 @@ namespace Opm
         return AutoDiffBlock<Scalar>::function( std::move(val), std::move(jac) );
     }
 
+    /**
+     * @brief Computes the value of base raised to the power of exp
+     *
+     * @param base The base AD forward block
+     * @param exp  The exponent AD forward block
+     * @return The value of base raised to the power of exp
+     */    template <typename Scalar>
```
Comment (atgeirr) [inline opm/autodiff/AutoDiffBlock.hpp]: Missing newline here.

File: opm/autodiff/AutoDiffBlock.hpp [hunk 2]
```diff
@@ -676,6 +677,50 @@ namespace Opm
...
+    AutoDiffBlock<Scalar> pow(const AutoDiffBlock<Scalar>& base,
+                                    const AutoDiffBlock<Scalar>& exp)
+    {
+        const int num_elem = base.value().size();
+        assert(exp.value().size() == num_elem);
```
Comment (atgeirr) [inline opm/autodiff/AutoDiffBlock.hpp]: AutoDiffBlock also has a size() method, you can just say exp.size() here.

File: opm/autodiff/AutoDiffBlock.hpp [hunk 3]
```diff
@@ -676,6 +677,50 @@ namespace Opm
...
+        for (int i = 0; i < num_elem; ++i) {
+            val[i] = std::pow(base.value()[i], exp.value()[i]);
+        }
+
+        // (f^g)' = f^g * ln(f) * g' + g * f^(g-1) * f'
+        typename AutoDiffBlock<Scalar>::V der1 = val;
+        for (int i = 0; i < num_elem; ++i) {
+            der1[i] *= std::log(base.value()[i]);
+        }
+        std::vector< typename AutoDiffBlock<Scalar>::M > jac1 (base.numBlocks());
```
Comment (atgeirr) [inline opm/autodiff/AutoDiffBlock.hpp]: This (and parts below) depends on `base.numBlocks() == exp.numBlocks()`. That should be `assert()`ed.

However, in the case when one of the arguments has `derivative().empty() == true` this function will fail even though it is actually a simpler case (one argument is a constant). This should be checked at the start of the function, one can call the already-existing pow() overload if exp is a constant. There seems to be no overload for the other case (constant base raised to variable exponent), that should be added, too.

**Comment by joakim-hove** [comment]
Maybe use another variable name than 'exp' - that symbol is nearly reserved
for the exponential function?

Or am I misunderstanding here?
7. mars 2016 09:26 skrev "Tor Harald Sandve" <notifications@github.com>
følgende:

**Comment by totto82** [comment]
exp is renamed to exponent and a method for base raised to variable exponent is added.

**Comment by atgeirr** [comment]
Looks good to me now, especially that the various combinations of variable and constant input are tested (except pow(constant, constant) but I think it will be ok). Will merge when Travis is done.

**Pull request merged**